core/src/test/scala/pureconfig/ProductHintSuite.scala

```diff
@@ -0,0 +1,170 @@
+package pureconfig
+
+import com.typesafe.config.{ ConfigFactory, ConfigObject }
+import org.scalatest.prop.GeneratorDrivenPropertyChecks
+import org.scalatest.{ EitherValues, FlatSpec, Matchers }
+import pureconfig.error.{ KeyNotFound, UnknownKey }
+import pureconfig.syntax._
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+class ProductHintSuite extends FlatSpec with ConfigConvertChecks with Matchers with EitherValues with GeneratorDrivenPropertyChecks {
+
+  behavior of "ProductHint"
+
+  case class ConfWithCamelCaseInner(thisIsAnInt: Int, thisIsAnotherInt: Int)
+  case class ConfWithCamelCase(camelCaseInt: Int, camelCaseString: String, camelCaseConf: ConfWithCamelCaseInner)
+
+  val confWithCamelCase = ConfWithCamelCase(1, "foobar", ConfWithCamelCaseInner(2, 3))
+
+  /** return all the keys in a `ConfigObject` */
+  def allKeys(configObject: ConfigObject): Set[String] = {
```

Comment:
If you convert `ConfigObject` to `Config`, you can list all entries of the config through the `entrySet()` method, so I think this can be simplified to:

```scala
configObject.toConfig().entrySet().asScala.flatMap(_.getKey.split('.')).toSet
```